dev-docs/services/ci/README.md

```diff
@@ -0,0 +1,114 @@
+
+- Check for validity
+  - The author must be a known developer (or else a commit SHA can be individually approved via the CI service UI)
+  - The PR must be against the `main` branch
+- Run tests
+  - The CI service will run tests against the PR
+  - The CI service will report the results of the tests back to GitHub
+
+### Running Tests
+
+The process of running tests goes like:
+
+- CI will generate a new batch and submit it against the production Hail Batch service using its own service credentials
+- The batch contains jobs which will:
+  - Clone the appropriate branch and build a new set of docker images for the updated services.
+  - Deploy the batch suite of k8s services into one of many CI-owned namespaces in the Hail Batch cluster
+    - These namespaces are called 
+  - Run a series of tests against the services
+  - Each test:
+    - Submits a specially crafted batch to the newly deployed batch namespace
+    - Checks that the results of running the batch are what we would expect
+- The CI service polls the batch which it submits for overall success or failure.
+- Once the batch has either succeeded or failed, CI uses that result to report status back to GitHub
+
+Examples of CI test runs can be seen by searching through the production batch log, as long as you have developer
+permissions. For example: `/batches?q=user+%3D+ci%0D%0Atest+%3D+1`
+
+#### CI Testing Timeline
+
+The image below shows the CI testing timeline:
+
+![CI Testing Timeline](ci-test.png)
+
+
+## Merging PRs to the `main` branch
+
+When a PR's status changes, it will trigger a hook which alerts the CI service. The CI service will 
+determine whether the PR is ready to be merged. If it is, the CI service will merge the pull request.
+
+Readiness is determined by github status checks. The following conditions must be met:
+
+- The PR must be against the `main` branch
+- The PR must have passed all tests (and there must be none outstanding)
+- The PR must have sufficient approvals
+
+## Deploying services to the live infrastructure
+
+When a PR is merged into the `main` branch, the CI service will:
+
+- Check for validity
+  - The author must be a known developer
+  - The PR must be against the `main` branch
+- Create a new batch job to:
+  - Build various components and service images
+  - Deploy to a test namespace
+  - Run tests against the services
+  - Deploy to the `default` (ie prod) namespace
+  - A handful of final actions
+    - eg rolling out artifact registry cleanup policies, amongst many other things
+
+Note: It's not actually quite as waterfall-y as this. In fact the jobs are all running in a hail
+batch, and each service being deploy has its own path through the DAG. So it's quite possible that the services are 
+test/deploy-ing in parallel, and that the deploy for one service might happen before the test for another has even begun.
+
+This should all be fine, because it was previously tested as part of the PR approval process.
```

Comment:
Do we really deploy anything before all test steps have succeeded? My mental model was that we ignore the tests run on the pr pre-merging, and treat it as untested post-merge, because the merge could have introduced semantic conflicts.

Reply:
This is a good question. This is an assumption based on - 
1. The batch lists the tasks in service-by-service order
2. The "deploy batch" job is listed before "test batch" jobs
3. There only seems to be one "deploy batch" job and it seems to target the `default` namespace

But I could easily have missed something. Worth double checking

Reply:
(Side-note: oh wow, some job groups would be a **_really!_** nice addition when trying to work this kind of thing out!)

Some selected examples from a [recent](https://batch.hail.is/batches/8199317) deploy:
- Job 27: `batch_image` [ran](https://batch.hail.is/batches/8199317/jobs/27) from 21:58 to 22:00
- Job 39: `test_hail_java_0` [ran](https://batch.hail.is/batches/8199317/jobs/39) from 22:12 to 22:17
- Job 44: `test_hail_python_0` [ran](https://batch.hail.is/batches/8199317/jobs/44) from 22:12 to 22:18
- Job 78: `test_hail_python_local_backend_0` [ran](https://batch.hail.is/batches/8199317/jobs/78) from 22:15 to 22:21
- Job 153 `deploy_batch` [ran](https://batch.hail.is/batches/8199317/jobs/153) from 22:04 to 22:05
- Job 176: `test_batch_0` [ran](https://batch.hail.is/batches/8199317/jobs/176) from 22:15 to 22:23
- Job 182: `deploy_ci` [ran](https://batch.hail.is/batches/8199317/jobs/182) from 22:05 to 22:05

Reply:
Note that you can see the dependencies directly in [build.yaml](https://github.com/hail-is/hail/blob/main/build.yaml). I believe `deploy_batch` is what deploys the new batch service in the test namespace, as `test_batch` depends on it. I'm not sure which job actually deploys to production.

Reply:
To close the loop here: `deploy_batch` does deploy to production (in the default namespace), which happens before tests are run. *But*, CI will only merge a PR if it has a successful CI run *on top of the current head of main*. That means the commit that will be merged and deployed is exactly the same (with the same git sha) as the commit that passed CI. There are still a few small holes for issues to slip past the tests, in particular because CI runs the batch tests in the test namespace in a fresh deployment, while the tests after deploying are run in the default namespace against the production system.